### PR TITLE
[MRG+2]: fixed 3804 and added test

### PR DESCRIPTION
**Pull request by ppasler**

### CODE REVIEW

**Comment by Eric89GXL** [comment]
If you write this in your description or a comment it will be linked (and auto-closed when this gets merged):

Closes #3804.

Incidentally I *think* if you had `#3804` instead of just `3804` in your commit message it might have linked.

**Comment by Eric89GXL** [comment]
Otherwise LGTM. Once you make my change, prefix your PR title with `MRG:` so that reviewers know that it's ready to be merged from your end

File: mne/preprocessing/tests/test_ica.py [hunk 1]
```diff
@@ -672,4 +672,14 @@ def test_eog_channel():
...
 
+def test_get_sources_repr():
+    """Test __repr__ from Raw object created by ica (test for #3804)"""
+    raw = read_raw_fif(raw_fname, preload=True).crop(0.5, stop).load_data()
+    n_components = 0.9
+    ica = ICA(n_components=n_components, method='fastica').fit(raw)
+    ica_raw = ica.get_sources(raw)
+    
+    assert_equal(ica_raw._filenames, [None])
+    print(repr(ica_raw))
```
Comment (dengemann) [inline mne/preprocessing/tests/test_ica.py]: Nitpick: actually calling repr or just printing is enough

**Comment by dengemann** [comment]
@ppasler since you fixed a bug, why don't you inform the world about this by adding an entry in the whatsnew file in the BUGS section for the next release?

**Comment by ppasler** [comment]
@dengemann sure, how do I do that? But is it really that important?

**Comment by dengemann** [comment]
> @dengemann sure, how do I do that? But is it really that important?

Yes it is, we need to inform our users about important code changes. Bug fixes belong to that class. If you don't want to do it I can do it this time. But you better get used to it early :)

File: doc/whats_new.rst [hunk 2]
```diff
@@ -1751,7 +1753,7 @@ Changelog
    - SourceEstimate now supports algebra by `Alex Gramfort`_.
 
 API changes summary
-~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (dengemann) [inline doc/whats_new.rst]: some unintended changes.

File: doc/whats_new.rst [hunk 3]
```diff
@@ -70,6 +70,8 @@ BUG
 
     - Fix a bug in :meth:`mne.time_frequency.AverageTFR.plot` when plotting without a colorbar by `Jaakko Leppakangas`_
 
+    - Fix ``_filenames`` attribute in creation of :class:`mne.io.RawArray` with :func:`mne.preporcessing.ica._sources_as_raw` by `Paul Pasler`_
+
```
Comment (dengemann) [inline doc/whats_new.rst]: take a look at the bottom of the file, there you can link  `Paul Pasler`_ to your webpage or github account.

File: doc/whats_new.rst [hunk 4]
```diff
@@ -1753,7 +1753,7 @@ Changelog
    - SourceEstimate now supports algebra by `Alex Gramfort`_.
 
 API changes summary
-~~~~~~~~~~~~~~~~~~~~~~~~~~~
+~~~~~~~~~~~~~~~~~~~
```
Comment (dengemann) [inline doc/whats_new.rst]: it seems you have removed a few characters inadvertently. We try to avoid that, even if it may have some good side-effects.

**Comment by dengemann** [comment]
@ppasler I will merge this once the build bots are happy.

**Comment by ppasler** [comment]
@dengemann alright - thanks, this was fun :)

File: mne/preprocessing/tests/test_ica.py [hunk 5]
```diff
@@ -672,4 +672,14 @@ def test_eog_channel():
...
 
+def test_get_sources_repr():
+    """Test __repr__ from Raw object created by ica (test for #3804)"""
+    raw = read_raw_fif(raw_fname, preload=True).crop(0.5, stop).load_data()
+    n_components = 0.9
+    ica = ICA(n_components=n_components, method='fastica').fit(raw)
+    ica_raw = ica.get_sources(raw)
+    
+    assert_equal(ica_raw._filenames, [None])
+    print(ica_raw)
```
Comment (agramfort) [inline mne/preprocessing/tests/test_ica.py]: can't we just update a test to avoid unnecessary test time?

Comment (dengemann) [inline mne/preprocessing/tests/test_ica.py]: Indeed. That would be good.

Comment (ppasler) [inline mne/preprocessing/tests/test_ica.py]: I added it to an existing test. 
But as I know unit testing, this method is already too long, as 1 test should only test 1 thing.

File: mne/preprocessing/tests/test_ica.py [hunk 6]
```diff
@@ -219,6 +219,10 @@ def test_ica_core():
         sources = ica.get_sources(raw)[:, :][0]
         assert_true(sources.shape[0] == ica.n_components_)
 
+        # test for #3804
+        assert_equal(sources._filenames, [None])
+        print(sources)
+
```
Comment (dengemann) [inline mne/preprocessing/tests/test_ica.py]: Yes, this is good.

File: doc/whats_new.rst (hunk 3 above)
Comment (dengemann) [inline doc/whats_new.rst]: Btw., back here, I think the wording for this message is not optimal. What you actually fix is the 'filenames' field in the measurement info.

Comment (ppasler) [inline doc/whats_new.rst]: So I should use field instead of attribute?

File: doc/whats_new.rst [hunk 7]
```diff
@@ -70,7 +70,7 @@ BUG
 
     - Fix a bug in :meth:`mne.time_frequency.AverageTFR.plot` when plotting without a colorbar by `Jaakko Leppakangas`_
 
-    - Fix ``_filenames`` attribute in creation of :class:`mne.io.RawArray` with :func:`mne.preporcessing.ica._sources_as_raw` by `Paul Pasler`_
+    - Fix the initialisation of the ``_filenames`` field while creation of :class:`mne.io.RawArray` in :func:`mne.preporcessing.ica._sources_as_raw` by `Paul Pasler`_
```
Comment (dengemann) [inline doc/whats_new.rst]: @ppasler damn it my bad, your initial description was better. I was too fast. Please revert the last commit.

**Comment by dengemann** [comment]
@ppasler you've got a failing test.
 
```
======================================================================
ERROR: Test ICA on raw and epochs.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\conda\envs\test\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "C:\projects\mne-python\mne\utils.py", line 819, in dec
    return function(*args, **kwargs)
  File "C:\projects\mne-python\mne\preprocessing\tests\test_ica.py", line 223, in test_ica_core
    assert_equal(sources._filenames, [None])
AttributeError: 'numpy.ndarray' object has no attribute '_filenames'
-------------------- >> begin captured stdout << ---------------------
<ICA  |  no decomposition, fit (fastica):  samples, no dimension reduction>
--------------------- >> end captured stdout << ----------------------
```

**Comment by jaeilepp** [comment]
You have a PEP issue
```
Running flake8

mne/preprocessing/tests/test_ica.py:223:1: W293 blank line contains whitespace

1

make: *** [flake] Error 1
```

**Comment by Eric89GXL** [comment]
Thanks for the fixes @ppasler, will merge once Travis comes back happy.

**Comment by Eric89GXL** [comment]
Actually there is one more test that needs to be fixed:

https://travis-ci.org/mne-tools/mne-python/jobs/180693943#L2815

Change it to `assert_equal(len(ica_raw._filenames), 0)` and you should be good. You can check locally by doing `nosetests mne//preprocessing/tests/test_ica.py`.

**Comment by ppasler** [comment]
@Eric89GXL actually this wasn't something I touched :) The line was changed by you two years ago.
But as the boyscout rule says:

>  Always leave the campground cleaner than you found it.

By the way, how can I run the static code analysis on my local machine?

**Comment by Eric89GXL** [comment]
Ahh crap I miswrote that -- I meant to say `assert_equal(..., 1)` because your code changes/fixes it to actually have an entry instead of no entry.

You can run `make flake` locally if on *NIX system, or just invoke `flake8` from the `mne-python` git root. You'll need `flake8` but it's conda and pip-installable.

**Comment by ppasler** [comment]
args, it seems the unit-test uses the installed version of mne... so I didn't notice the failure.

**Comment by Eric89GXL** [comment]
Travis OSX was taking forever to start so I killed it. This one should be good to go. Thanks @ppasler

**Pull request merged**